### DISCUSSION TITLE
How does PyVista relate to other visualization tools?

### DISCUSSION THREAD

**Discussion by banesullivan** [pyvista/pyvista#1438] [Ideas]
I think it might be useful to get a comparison table going in the docs to compare `pyvista` to other 3D viz tools out there (Python-based or not). 

Let's use this issue as a place to keep track of all the different tools out there start outlining what sections might be included in the comparison table. 

### Example

Here's an example of what I'm imaging for this:

While this table outlines the various features available across these tools, it doesn't capture the ease-of-use of those features across the different software packages or other factors such as ease of installation or the *Pythonic-ness* of the API (how intuitive the API is to a new user)

|                                    | PyVista | `plotly` | Mayavi | `ipyvolume` | `vtkplotter` | ParaView |
|------------------------------------|--------|----------|--------|-------------|--------------|----------|
| Python API                         | ✅      | ✅        | ✅      | ✅           | ✅            | ✅        |
| `pip` installable                  | ✅      | ✅        | ✅      | ✅           | ✅            | ❌        |
| Free & Open-Source                 | ✅      | ✅        | ✅      | ✅           | ✅            | ✅        |
| Supports 3D                        | ✅      | ✅        | ✅      | ✅           | ✅            | ✅        |
| Supports Time                      | ✅      | ✅        | ✅      | ✅           | ✅            | ✅        |
| NumPy Interface                    | ✅      | ✅        | ✅      | ✅           | ✅            | ❌        |
| Supports Large Data                | ✅      | ❌        | ✅      | ❌           | ✅            | ✅        |
| Jupyter Notebooks                  | ✅      | ✅        | ✅      | ✅           | ✅           | ❌        |
| Exports to VTKjs or X3D            | ✅      | ❌        | ✅      | ❌           | ✅            | ✅        |
| Eye-Dome Lighting (EDL)            | ✅      | ❌        | ❌      | ❌           | ❌            | ✅        |
| Focuses on Finite Element/Volume Types                | ✅      | ❌        | ✅      | ❌           | ❌           | ✅        |
| Analysis/Filtering Tools           | ✅      | ❌        | ✅      | ❌           | ✅            | ✅        |
| Opacity/Transparency               | ✅      | ✅        | ✅      | ✅           | ✅            | ✅        |
| Subplotting                        | ✅      | ❌        | ✅      | ❌           | ✅            | ✅        |
| Exports to movie/GIF               | ✅      | ✅        | ✅      | ✅           | ✅            | ✅        |
| Supports geometric glyphing        | ✅      | ❌        | ✅      | ✅           | ✅            | ✅        |
| Supports a variety of file formats | ✅      | ❌        | ✅      | ❌           | ✅            | ✅        |
| Direct access to filtering methods | ✅      | ❌        | ❌      | ❌           | ✅             | --        |
| Interactivity/embeddable in Jupyter notebooks | ✅ | ✅| ✅ | ✅ | ✅   | --        |

**Comment by banesullivan** [comment]
Should tools like ParaView, VisIt, and Leapfrog be on here too?

**Comment by akaszynski** [comment]
I’d add paraview and vtkplotter as well.

**Comment by banesullivan** [comment]
We should also get more items on the Highlights section of the documentation/README to showcase all of what `pyvista` can do! @akaszynski, do you have a list of notable features you find most useful?

**Comment by akaszynski** [comment]
The features I regularly used ``pyvista`` for:
- Organizing unstructured and structured grids from CAE software
  - Store point and cell scalars that are passed down to child meshes.
  - Seamless mesh copying and regeneration. 
- Visualizing CAE results.
  - Deformation animation
  - Mode shape plotting
  - Stress/strain visualization
- Surface extraction.
  - Surface mesh (triangular and polygonal) normal calculation
  - Surface area calculation
  - Mesh quality (based on a variety of parameters)
- Direct access to numpy arrays to allow for grid and surface mesh modification.

As you can see, this is mostly CAE (finite element analysis) based.  We also used vtki for the mesh processing backbone for another software package, femorph (see [press release](https://www.wpafb.af.mil/News/Article-Display/Article/1503043/afrl-signs-first-of-its-kind-software-license-with-pratt-whitney/)).  I'd have to write many functions from scratch to get all the capabilities that ``vtk`` offers, but as the interface isn't "pythonic" I ended up writing ``pyvista`` to make it easier to plug ``vtk`` into the workflow.

**Comment by larsoner** [comment]
Just to add from the perspective from a long-time mayavi+tvtk user (e.g., [PySurfer](https://pysurfer.github.io/auto_examples/index.html)) who is excited about the potential of this project (!):

- it looks like the overlap between vtki and mayavi+tvtk seems the most substantial (pythonic interface; easy NumPy support/integration). I'm particularly curious if there are maintainability/goal/scope/whatever differences between the libraries. This might be out of scope for these docs, but for folks like me who have endured Mayavi for years, you might be able to nudge them into adopting vtki with some arguments beyond the feature list, since it's pretty close to parity.
- FWIW in PySurfer were have "subplotting" grids using Mayavi the way they suggest in the docs, namely by setting up a proper traitsui window with multiple Mayavi scenes in it. So if you also accomplish subplotting by having multiple VTK scenes, then having an X in the Mayavi column there is probably not correct. But if there is another method that VTK supports natively that Mayavi does not, then the X probably makes sense (maybe with an asterisk, or with renaming to "native subplotting" or something).

**Comment by banesullivan** [comment]
Thanks, @larsoner! I wasn't aware that Mayavi could provide subplots as it sounds like we do a similar thing - I've updated the table to reflect this.

> it looks like the overlap between vtki and mayavi+tvtk seems the most substantial

You're definitely right that there is a lot of overlap in features between Mayavi and `pyvista`! I do however think `pyvista` is approaching 3D viz in a totally different way the Mayavi... first, `pyvista` is truly an interface to the Visualization Toolkit. We provide an easy to use interface to VTK's Python bindings making accessing VTK data objects simple and fast. This allows `pyvista` to merge into any existing Python VTK code as `pyvista` objects are instances of VTK objects. It also stays true to VTK's object-oriented approach.

For example, in `pyvista` we simply wrap common VTK classes with properties and methods to make accessing the underlying data within the VTK data object and using VTK filters more intuitive so that users don't need to know the nuances of creating VTK pipelines and remember all the different VTK classes for filters, etc. I think Mayavi has a similar effort in this regard but I don't know enough to comment too much further. I do know that the differences bewteen how `pyvista` and Mayavi make VTK filters available to the user are stark:

```py
from pyvista import examples
dataset = examples.load_uniform()

# Apply VTK threshold filter
result = dataset.threshold([100, 300])
```
vs. in Mayavi a user has to search around in the `mlab.pipeline` module for what their looking for and set up a pipeline. `pyvista` just makes any filter that a given data type can support directly available on that object as a callable method with keyword arguments. This makes rapid prototyping in IPython a breeze with tab-complete.

> I'm particularly curious if there are maintainability/goal/scope/whatever differences between the libraries.

I'll work towards outlining these more clearly! At the moment, my primary goal with `pyvista` is to make it incredibly intuitive to use - thus making 3D visualization more accessible to a broad audience (particularly geoscientists). I've dabbled with Mayavi a few times and found the docs incredibly difficult to follow - and I hear from other users its difficult to install... so our goal with `pyvista` is to break down the entrance barrier for new Python users to start creating compelling 3D visualizations. 

Also, since `pyvista` is pure Python and VTK has wheels on PyPI, using `pyvista` as a supporting module for 3D visualization is really easy for external projects to harness. For example, check out what's possible in the [connections](http://docs.pyvista.org/#connections) section of the docs.

To be frank, the few times I've dived into using Mayavi, I found it cumbersome and not intuitive - perhaps I should create a docs page comparing examples between `pyvista` and Mayavi to demonstrate the benefit of one library over the other.

**Comment by larsoner** [comment]
> in Mayavi a user has to search around in the mlab.pipeline module for what their looking for and set up a pipeline... my primary goal with vtki is to make it incredibly intuitive to use... I've dabbled with Mayavi a few times and found the docs incredibly difficult to follow - and I hear from other users its difficult to install... so our goal with vtki is to break down the entrance barrier for new Python users to start creating compelling 3D visualizations... To be frank, the few times I've dived into using Mayavi, I found it cumbersome and not intuitive

Agreed on all fronts here :) Saying that the goal is to provide a pythonic, well documented interface to expose all VTK functionality would be a good start already. I remember it took me some number of hours to get the "terrain" (instead of the default "trackball") interaction stuff working in Mayavi. In the end the code was simple, but the path to get there was not...

> perhaps I should create a docs page comparing examples between vtki and Mayavi to demonstrate the benefit of one library over the other.

This would be excellent. Maybe when we figure out how to embed interactive 3D visualizations using Javascript in sphinx-gallery (someday, somehow!) I'll simultaneously adapt PySurfer to have a vtki backend. I don't expect adding vtki as a backend to be too much work, and am optimistic that the vtki code will indeed end up more readable :)

One example that was painful for me to come up with, if you're looking for ideas, was [plotting arbitrary RGBA values on a triangulated mesh](https://pysurfer.github.io/auto_examples/plot_custom_colors.html#sphx-glr-auto-examples-plot-custom-colors-py). It's not very readable.

> since vtki is pure Python and VTK has wheels on PyPI, using vtki as a supporting module for 3D visualization is really easy for external projects to harness.

FWIW `mayavi+tvtk` is pure python plus one tiny compiled extension and recently (last 6 mo?) made the compiled extension optional. So we no longer really experience `pip install mayavi` problems on Linux/OSX/Windows, beyond any brought on by the `pip install vtk` itself (which have also mercifully been few in the last year or so).

**Comment by banesullivan** [comment]
> Saying that the goal is to provide a pythonic, well documented interface to expose all VTK functionality would be a good start already.

That's exactly our mission with `pyvista`!! 

> I remember it took me some number of hours to get the "terrain" (instead of the default "trackball") interaction stuff working in Mayavi. In the end the code was simple, but the path to get there was not...

It's quite easy in `pyvista` with [`Plotter.enable_terrain_style()`](http://docs.vtki.org/en/latest/tools/plotting.html#vtki.BasePlotter.enable_terrain_style) 😊

```py
import pyvista as pv
from pyvista import examples

# Load St Helens DEM and warp the topography
mesh = examples.download_st_helens().warp_by_scalar()

p = pv.Plotter(notebook=False)
p.add_mesh(mesh)
p.enable_terrain_style()
p.show()
```
>One example that was painful for me to come up with, if you're looking for ideas, was plotting arbitrary RGBA values on a triangulated mesh. It's not very readable.

I added the ability to plot RGB values not too long ago in #114 - I could expand this to support transparency as well. Could you send me the data file needed for that example? I install `pysurfer` and downloaded that example but have an issue because of `'SUBJECTS_DIR'`.

If I have that data file... the code to make the RGB(A) plot (A when I add transparency) would be:

```py
# !pip install pysurfer
from surfer import Brain
import pyvista as pv
import numpy as np

b = Brain('fsaverage', hemi, surf, subjects_dir=subjects_dir,)

# co-ordinates
verts = b.geo[hemi].coords.T
tris = b.geo[hemi].faces

brain = pv.Poly(verts, tris)
brain.plot(rgb=True)
```

**Comment by larsoner** [comment]
http://faculty.washington.edu/larsoner/fsaverage_min.zip

If you download this file and extract it to some directory, set that directory to the `SUBJECTS_DIR` environment variable. (For example, extract to `/home/larsoner/temp` such that the directory `/home/larsoner/temp/fsaverage/label` now exists, and `export SUBJECTS_DIR=/home/larsoner/temp`.) Then hopefully the example will run for you.

**Comment by banesullivan** [comment]
Is there a way to create the Brain object without creating an mlab figure? It's non-responsive in a Jupyter notebook for me...

**Comment by larsoner** [comment]
No but you can load the surface geometry with
```
>>> from surfer import Surface
>>> surf = Surface('fsaverage', 'lh', 'inflated')
>>> surf.load_geometry()
>>> surf.coords
array([[ -5.88037491,   2.4935925 ,  57.62179947],
       [ 16.4801445 , -64.23391724,  54.90657043],
       [ 30.06188965,  16.55784416,  43.14960861],
       ...,
       [ -6.03076267, -10.78562832, -61.8919487 ],
       [ -6.26786232, -10.91573811, -61.979599  ],
       [ -6.49331236, -11.02756691, -62.06248474]])
>>> surf.faces
array([[     0,  40964,  40962],
       [     0,  40962,  40965],
       [     0,  40965,  40967],
       ...,
       [163841, 160926,  39990],
       [163113, 160926, 163841],
       [163113,     11, 160926]], dtype=int32)
```

**Comment by banesullivan** [comment]
FYI, see 1580f3d which enabled RGBA plotting!

Here's an example if you install the master branch of `pyvista`:

```bash
pip install -U git+git://github.com/pyvista/pyvista@master
```

```py
import numpy as np
import matplotlib.pyplot as plt
import pyvista as pv
from surfer import Surface

# Load data from pysurfer
sdir = '/Users/bane/Desktop/viz-projects/'
hemi = 'lh'
surf = 'white'
surf = Surface('fsaverage', hemi, surf, subjects_dir=sdir)
surf.load_geometry()

verts = surf.coords
tris = surf.faces

# generate an rgba matrix, of shape n_vertices x 4

def norm(x):
    '''Normalise array betweeen 0-1'''
    return (x - np.min(x)) / (np.max(x) - np.min(x))

# define color map
cmap = plt.cm.viridis

# change colour based on position on the x axis
hue = norm(verts[:,0])
colors = cmap(hue)[:, :3]

# change alpha based on position on the z axis
alpha = norm(verts[:,-1])

# combine hue and alpha into a Nx4 matrix
rgba_vals = np.concatenate((colors, alpha[:, None]), axis=1)
rgba_vals = (rgba_vals * 255).astype('ubyte')

##################### pyvista code ######################

# Define cell type for vtk: 3 = triangle
faces = np.c_[np.full(len(tris), 3, dtype=np.int32), tris]

# Create the pyvista object and render
brain = pv.PolyData(verts, faces)

# And render it
brain.plot(scalars=rgba_vals, rgba=True,)# notebook=False)
```
![download](https://user-images.githubusercontent.com/22067021/55501487-bb618100-5607-11e9-900a-8e0b57012467.png)

**Comment by larsoner** [comment]
Nice!

**Comment by banesullivan** [comment]
- [ ] Another library to add to the table: https://github.com/vispy/vispy

**Comment by nicoguaro** [comment]
- [ ] Another library to add is [yt](https://yt-project.org/).

**Comment by banesullivan** [comment]
Thanks for pointing out `yt`, @nicoguaro!

I was under the impression that while `yt` claims to be a "package for analyzing and visualizing volumetric data," their plotting remained 2D. From glancing over their documentation website a bit deeper this time, I'm realizing they do indeed support 3D rendering.

`yt` seems to be a powerful library and I'll be sure to include it in our comparison table.

**Comment by banesullivan** [comment]
if you @nicoguaro (or anyone) are more familiar with `yt`, perhaps you could help us fill out ✅s and ❌s for each of the categories in the table for `yt`?

**Comment by larsoner** [comment]
@banesullivan it's probably worth having a row entry for *interactivity with embedded Jupyter notebook plot* (not static PNG, not via separate window), since it's a cool feature. It looks like Mayavi can more or less do it (see https://github.com/nipy/PySurfer/pull/268):

https://mybinder.org/v2/gh/wmvanvliet/SNL_workshop_2019/master?filepath=pysurfer_demo.ipynb

So it's probably also doable for pyvista (or maybe it is already)?

**Comment by banesullivan** [comment]
> So it's probably also doable for pyvista (or maybe it is already)?

It is already! Check out all the examples on http://playground.pyvista.org

Or this one: https://mybinder.org/v2/gh/OpenGeoVis/omfvista/master?filepath=Example.ipynb

![omfvtk](https://user-images.githubusercontent.com/22067021/58102911-14817600-7bb0-11e9-8be0-6f463b251241.gif)

**Comment by larsoner** [comment]
Excellent!

**Comment by banesullivan** [comment]
Just added that category to the table above, thanks for the suggestion @larsoner

**Comment by marcomusy** [comment]
..actually i think vtkplotter has all those functionalities except EDL :)

**Comment by banesullivan** [comment]
Thanks, @marcomusy! I updated the table to reflect `vtkplotter`'s capabilities

**Comment by banesullivan** [comment]
One thought on the "Explicit Data Types" category - maybe we should rephrase this to reflect the differences in PyVista's types, `vtkplotter`'s types, whatever Mayavi uses, and `plotly`'s lack of types. I'm thinking this because there is a major difference in the underlying structures between at least PyVista and `vtkplotter` since PyVista wraps finite element/volume meshes/volumes while `vtkplotter` wraps actors and not necessarily the underlying data types.

Also, we want to highlight PyVista's advantages here and showcase that PyVista can be used for mesh analysis, refinement, processing, modelling, and so much more than just 3D visualization

**Comment by selvakarna** [comment]
How to export surface/mesh into Cad support IGES/STEP File?

**Comment by banesullivan** [comment]
Hi @selvakarna, it sounds like you have a usage question off topic from this issue. Would you please make a new issue with your question and provide details/documentation/links for the file format you mention?

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by tkoyama010** [comment]
@banesullivan  This table is great. Can I use this in my talk presentation?

**Comment by nicoguaro** [comment]
Should ``conda installable`` a new row in the table?

**Comment by banesullivan** [comment]
@tkoyama010, feel free to use, but with caution (this hasn't been updated in a while).

@nicoguaro, that's a good idea! I'll have to come back to this soon, update the table, then finally close this issue and but the table in the docs

**Comment by rcameronc** [comment]
Hi all, I just discovered PyVista and am very impressed with it!  One question: I am trying to build 3D models of ice sheets that cover continents, which means I need to project them (i.e. WGS 84).  I see a PyVista example for 2D data in spherical coordinates -- how would you suggest extending that to 3D topography on a sphere?

**Comment by banesullivan** [comment]
Hi there @rcameronc, please open a support issue here: https://github.com/pyvista/pyvista-support

It will help to include sample data and further context. This thread is focused on comparing PyVista to other 3D viz tools at a high level.
